pkg/syncer/storage/storage_pvc_controller.go

```diff
@@ -0,0 +1,200 @@
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/logging"
+	"github.com/kcp-dev/kcp/third_party/keyfunctions"
+)
+
+const persistentVolumeClaimControllerName = "kcp-workload-syncer-storage-pvc"
+
+var (
+	persistentVolumeClaimSchemeGroupVersion = corev1.SchemeGroupVersion.WithResource("persistentvolumesclaims")
+	persistentVolumeSchemeGroupVersion      = corev1.SchemeGroupVersion.WithResource("persistentvolumes")
+)
+
+type PersistentVolumeClaimController struct {
+	queue                                 workqueue.RateLimitingInterface
+	upstreamInformers                     kcpdynamicinformer.DynamicSharedInformerFactory
+	downstreamClient                      dynamic.Interface
+	syncTarget                            syncTargetSpec
+	updateDownstreamPersistentVolumeClaim func(ctx context.Context, persistentVolumeClaim *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error)
+	updateDownstreamPersistentVolume      func(ctx context.Context, persistentVolume *corev1.PersistentVolume) (*corev1.PersistentVolume, error)
+	getDownstreamPersistentVolumeClaim    func(persistentVolumeClaimName, persistentVolumeClaimNamespace string) (runtime.Object, error)
+	getDownstreamPersistentVolume         func(persistentVolumeName string) (runtime.Object, error)
+	getUpstreamPersistentVolumeClaim      func(clusterName logicalcluster.Name, persistentVolumeClaimName, persistentVolumeClaimNamespace string) (runtime.Object, error)
+}
+
+// syncTargetSpec contains all the details about a given sync target.
+type syncTargetSpec struct {
+	name      string
+	workspace logicalcluster.Name
+	uid       types.UID
+	key       string
+}
+
+// NewPersistentVolumeClaimSyncer returns a new storage persistent volume claim syncer controller.
+func NewPersistentVolumeClaimSyncer(
+	syncerLogger logr.Logger,
+	syncTargetWorkspace logicalcluster.Name,
+	syncTargetName, syncTargetKey string,
+	downstreamClient dynamic.Interface,
+	downstreamKubeClient *kubernetes.Clientset,
+	upstreamInformers kcpdynamicinformer.DynamicSharedInformerFactory,
+	downstreamInformers dynamicinformer.DynamicSharedInformerFactory,
+	syncTargetUID types.UID,
+) (*PersistentVolumeClaimController, error) {
+
+	c := &PersistentVolumeClaimController{
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), persistentVolumeClaimControllerName),
+		downstreamClient:  downstreamClient,
+		upstreamInformers: upstreamInformers,
+		syncTarget: syncTargetSpec{
+			name:      syncTargetName,
+			workspace: syncTargetWorkspace,
+			uid:       syncTargetUID,
+			key:       syncTargetKey,
+		},
+		updateDownstreamPersistentVolumeClaim: func(ctx context.Context, persistentVolumeClaim *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+			persistentVolumeClaim, err := downstreamKubeClient.CoreV1().PersistentVolumeClaims(persistentVolumeClaim.Namespace).Update(ctx, persistentVolumeClaim, metav1.UpdateOptions{})
+			if err != nil {
```

Comment:
nit: you can just return the values from the client call here, the `if` logic does not change anything